### PR TITLE
Added scroll bar to ListBox2D

### PR DESCRIPTION
**Pull request by karandeepSJ**
This PR adds a scroll bar to the `ListBox2D` in viz.ui. The scroll bar can be moved at different speeds with the mouse. It also moves up or down when the scrolling occurs through the mouse wheel.
scroll

### CODE REVIEW

**Comment by skoudoro** [comment]
Hi @karandeepSJ, Thanks for this cool features!  Can you check [this error](https://travis-ci.org/nipy/dipy/jobs/392114901#L4951) on travis?

File: dipy/viz/tests/test_ui.py [hunk 1]
```diff
@@ -528,41 +528,8 @@ def _on_change():
...
-
-    # Test without multiselection enabled.
-    listbox.multiselection = False
-    del selected_values[:]  # Clear the list.
-    show_manager.play_events_from_file(recording_filename)
-
-    # Check if the right values were selected.
-    expected = [[1], [2], [2], [21], [1], [42]]
-    assert len(selected_values) == len(expected)
-    assert_arrays_equal(selected_values, expected)
```
Comment (skoudoro) [inline dipy/viz/tests/test_ui.py]: Why are you removing this test?

**Comment by Garyfallidis** [comment]
Same here @karandeepSJ . Needs rebasing.

**Comment by Garyfallidis** [comment]
@karandeepSJ please resolve the conflicts of this PR.

**Comment by dmreagan** [comment]
This looks pretty nice, but I've observed some small quirks in the scrollbar behavior. I found that if you drag the bar downward in small motions, it scrolls up instead of down. The reverse is also true. I could only trigger this behavior when the scrollbar is large.

scrollbar_bug

**Comment by dmreagan** [comment]
The bug is fixed, so this PR looks good to me. How difficult would it be to change the color of the bar when it is grabbed like we do with sliders? It could wait for another PR. Any thoughts here, @ranveeraggarwal and @MarcCote ?

File: dipy/viz/tests/test_ui.py [hunk 2]
```diff
@@ -539,13 +539,38 @@ def test_ui_range_slider(interactive=False):
 
 @npt.dec.skipif(not have_vtk or skip_it)
 @xvfb_it
-def test_ui_listbox_2d(recording=False):
+def test_ui_listbox_2d(interactive=False):
```
Comment (MarcCote) [inline dipy/viz/tests/test_ui.py]: Like we discussed in PR #1559 , we may want to have a `mode` parameter that can either accep `None, `interactive` or `recording`.

File: dipy/viz/ui.py [hunk 3]
```diff
@@ -3009,41 +3014,57 @@ def _setup(self):
...
-        self.up_button = Button2D([("up", arrow_up)])
-        pos = self.panel.size - self.up_button.size // 2 - margin
-        self.panel.add_element(self.up_button, pos, anchor="center")
-
-        arrow_down = read_viz_icons(fname="arrow-down.png")
-        self.down_button = Button2D([("down", arrow_down)])
-        pos = (pos[0], self.up_button.size[1] // 2 + margin)
-        self.panel.add_element(self.down_button, pos, anchor="center")
+        # arrow_up = read_viz_icons(fname="arrow-up.png")
+        # self.up_button = Button2D([("up", arrow_up)])
```
Comment (MarcCote) [inline dipy/viz/ui.py]: Oh, let's remove those commented lines.

File: dipy/viz/ui.py [hunk 4]
```diff
@@ -3009,41 +3014,57 @@ def _setup(self):
...
+                                 size=(slot_width, self.slot_height))
             item.textblock.font_size = font_size
             item.textblock.color = (0, 0, 0)
             self.slots.append(item)
             self.panel.add_element(item, (x, y + margin))
 
         # Add default events listener for this UI component.
-        self.up_button.on_left_mouse_button_pressed = self.up_button_callback
-        self.down_button.on_left_mouse_button_pressed = self.down_button_callback
+        # self.up_button.on_left_mouse_button_pressed = self.up_button_callback
```
Comment (MarcCote) [inline dipy/viz/ui.py]: Remove commented lines.

**Comment by karandeepSJ** [comment]
@dmreagan  This is how it looks like when the color of the scroll bar changes on grabbing it. Currently, I have only made it darker.

scroll_color

**Comment by karandeepSJ** [comment]
@dmreagan  @skoudoro  @Garyfallidis There were some connection problem during the travis build of this PR and of #1559.

**Comment by dmreagan** [comment]
@karandeepSJ please rebase

**Comment by karandeepSJ** [comment]
Rebased. To adjust the scroll bar size and step length on changing a directory, I added a function `update_scrollbar` to `ListBox2D'. I also added the `update_element` function to `Panel2D` to update the position of the scroll bar in the panel

**Comment by dmreagan** [comment]
@MarcCote please take a look at @karandeepSJ's latest update. It looks like `update_element` is repeating a lot of code from `add_element`. Is there a better way to do this?

**Comment by MarcCote** [comment]
@dmreagan @karandeepSJ we could refactor the `update_element` code like this
```python

    
    def remove_element(self, element):
        idx = self._elements.index(element)
        del self._elements[idx]
        del self.element_offsets[idx]

    def update_element(self, element, coords, anchor="position"):
        """ Updates the position of a UI component in the panel.

        Parameters
        ----------
        element : UI
            The UI item to be updated.
        coords : (float, float) or (int, int)
            New coordinates.
            If float, normalized coordinates are assumed and they must be
            between [0,1].
            If int, pixels coordinates are assumed and it must fit within the
            panel's size.
        """
        self.remove_element(element)
        self.add_element(element, coords, anchor)
```

**Comment by skoudoro** [comment]
Hi @karandeepSJ, Any chance to have the improvement above proposed by @MarcCote  before merging this PR?

**Comment by karandeepSJ** [comment]
Yes, I am working on it

**Comment by karandeepSJ** [comment]
Done. Also, there was an error being thrown when I tried to open a directory without read permissions from the filemenu. I got rid of the error. Now the user just stays in the same directory

**Comment by skoudoro** [comment]
Great! It is ready to be merged. Can you do it @dmreagan? 

Thank you @karandeepSJ!

**Pull request merged**